src/runtime_src/core/tools/xbutil2/OO_HostMem.cpp

```diff
@@ -0,0 +1,169 @@
+  else
+    throw xrt_core::generic_error(EINVAL, "Invalid host-mem action '" + str + "'");
+}
+
+void
+host_mem(xrt_core::device* device, bool action, uint64_t size)
+{
+  XBUtilities::sudo_or_throw("Root privileges required to enable host-mem");
+  device->set_cma(action, size); //can throw
+}
+
+} //end namespace 
+
+// ----- C L A S S   M E T H O D S -------------------------------------------
+
+OO_HostMem::OO_HostMem( const std::string &_longName, bool _isHidden )
+    : OptionOptions(_longName, _isHidden, "Controls host-mem functionality")
+    , m_devices({})
+    , m_action("")
+    , m_size("")
+    , m_help(false)
+{
+  m_optionsDescription.add_options()
+    ("device,d", boost::program_options::value<decltype(m_devices)>(&m_devices)->multitoken(), "The Bus:Device.Function (e.g., 0000:d8:00.0) device of interest")
+    ("action", boost::program_options::value<decltype(m_action)>(&m_action)->required(), "Action to perform: ENABLE or DISABLE")
+    ("size,s", boost::program_options::value<decltype(m_size)>(&m_size), "Size of host memory to be enabled (e.g. 256M, 1G)")
+    ("help,h", boost::program_options::bool_switch(&m_help), "Help to use this sub-command")
+  ;
+
+  m_positionalOptions.
+    add("action", 1 /* max_count */)
+  ;
+}
+
+void
+OO_HostMem::execute(const SubCmdOptions& _options) const
+{
+  XBUtilities::verbose("SubCommand option: Host Mem");
+
+  XBUtilities::verbose("Option(s):");
+  for (auto & aString : _options)
+    XBUtilities::verbose(std::string(" ") + aString);
+
+  // Honor help option first
+  if (std::find(_options.begin(), _options.end(), "--help") != _options.end()) {
+    printHelp();
+    return;
+  }
+
+  // Parse sub-command ...
+  po::variables_map vm;
+
+  try {
+    po::store(po::command_line_parser(_options).options(m_optionsDescription).positional(m_positionalOptions).run(), vm);
+    po::notify(vm); // Can throw
+  }
+  catch (po::error& e) {
+    std::cerr << "ERROR: " << e.what() << "\n\n";
+
+    printHelp();
+    return;
+  }
+
+  // Exit if neither action or device specified
+  if(m_help || (m_action.empty() || m_devices.empty())) {
+    printHelp();
+    return;
+  }
+
+  bool enable = false;
+  try {
+    uint64_t size = 0;
+    if(!m_size.empty()) {
+      uint64_t bytes = 0;
+      auto units = m_size.substr(m_size.length()-1);
+      boost::to_upper(units);
+      if(units.compare("M") == 0)
+        bytes = 1024*1024;
+      else if(units.compare("G") ==0)
+        bytes = 1024*1024*1024;
+      else
+        throw xrt_core::error(std::errc::invalid_argument, "Please specify a valid size unit (M|G), eg: `256M`");
+
+      try {
+        size = std::stoll(m_size.substr(0, m_size.length()-1));
+      } 
+      catch (const std::exception& ex) {
+        //out of range, invalid argument ex
+        throw xrt_core::error(std::errc::invalid_argument, "Value supplied to --size option is invalid");
+      }
+      size = size*bytes;
+    }
+
+    enable = string2action(m_action);
```

Comment:
Being that this is a boolean variable, it would have been simpler to use the following code:
```
if ((boost::iequals(m_action, "ENABLE") != 0) && (boost::iequals(m_action, "DISABLE") != 0) {
   std::cerr << boost::format("ERROR: Invalid action value: '%s'\n) % m_action;
   printHelp();
   throw xrt_core::error(std::errc::operation_canceled);
 }

bool enable =  boost::iequals(m_action, "ENABLE");

```

Reply:
Hmm. I am a little confused...
Is it better to have 2-3 logics in the same if? In one of the previous comments, you asked me to separate each logic out.

Reply:
The first half of sample code  validates if the value is correct.  If not an error occurs and lets the user know what has happen.

After that we know that the value is either "ENABLE" or "DISABLE" and we use the iequals to determine it.   

The flow here is:
1. Validate the values.
2. Determine what the correct value is.

Note: This is preferred for a boolean value.